### ISSUE TITLE
Support for multiline strings in JSON

### ISSUE THREAD

**Issue by eduardomourar** [microsoft/vscode-azurearmtools#133]
The following is a valid ARM template:
```json
"[json( concat( '{',
   '\"element1\": \"', param( 'element1' ), '\",',
   '\"element2\": \"', param( 'element2' ), '\"',
'}' ) )]"
```
But VSCode presents that as an error because it is not a valid JSON. We would like to have that type error ignored for ARM templates.
<img width="540" alt="vscode_error" src="https://user-images.githubusercontent.com/16357187/46617408-ff723e00-cb1c-11e8-810b-42d2206cf3a9.png">

I can create a PR if someone can point me in the right direction.

**Comment by StephenWeatherford** [comment]
Can you point me to documentation on this?  I can't seem to find any.  Thx.

This might be problematic because some of the errors will come from vscode's jsonc functionality, not just from the extension.

**Comment by eduardomourar** [comment]
I also could not find anything in the ARM template documentation, but I know from experience (trial and error) that during Azure deployment templates are interpreted as single line (so new line within string are converted to `\n` by parser). You can check a user feedback [here](https://feedback.azure.com/forums/34192--general-feedback/suggestions/35082124-line-breaks-issue-with-arm-templates).

Just like you mentioned, the error does come from both json and jsonc functionality from VSCode (actual line [here](https://github.com/Microsoft/vscode-json-languageservice/blob/29c1edd535f76f5f2ae00f421248f5c6eea02737/src/parser/jsonParser.ts#L1053)).  This is the actual error in VSCode:
```
[json]  Unexpected end of string.
```
I guess we need to override that error just for ARM templates and not for the rest (because this is not a valid JSON). This is the reason I created the issue here and not in VSCode JSON repo.

**Comment by eduardomourar** [comment]
@StephenWeatherford , any news on this issue?

**Comment by StephenWeatherford** [comment]
I'm afraid I haven't had time to look at this yet.  If someone wants to step up and give it a try?

**Comment by StephenWeatherford** [comment]
@bmoore-msft not sure what the official stance is on this.
@eduardomourar I think this is going to be problematic. We don't have our own template language server - we depend on the existing JSON language server in vscode.  Since this isn't valid JSON, vscode is going to complain no matter what we do.

**Comment by bmoore-msft** [comment]
ARM will accept this, and some (maybe many I haven't found one that doesn't) JSON parsers will also ignore whitespace when loading the doc.

ACK that this is likely something VS Code needs to handle in their parser.

**Comment by eduardomourar** [comment]
@StephenWeatherford  and @bmoore-msft  thanks for the info. My original idea was to just override the error being thrown from the existing JSON language server, if possible. I do believe that there are more exceptions for ARM templates that can be added to the mix.

**Comment by bmoore-msft** [comment]
Understood - the challenge with simply "overriding" is that the ARM Tools are an "extension" to the JSON language server.  Extensions can generally only add functionality from the language server not remove it.  So it may not even be possible with the current architecture.  Stephen will know for sure.

**Comment by StephenWeatherford** [comment]
Not sure how easy this would be, or if az deploy would support it, etc.  Either way, the JSON language server will mark it as invalid JSON.

**Comment by StephenWeatherford** [comment]
The extension now has support for multi-line strings.  Unfortunately, the built-in JSON language server does not, so you still get errors from it.  We're working on improving that experience.

**Comment by kstro21** [comment]
What about supporting applying all of the validations, colorization, IntelliSense, etc. to  `YAML` files that we can later convert to JSON and deploy? That would be great.

**Comment by StephenWeatherford** [comment]
Fixed by #263! (releasing very soon)
